src/main/java/com/conveyal/datatools/manager/models/transform/PreserveCustomFieldsTransformation.java

```diff
@@ -0,0 +1,136 @@
+import com.conveyal.datatools.manager.models.TableTransformResult;
+import com.conveyal.datatools.manager.models.TransformType;
+import org.supercsv.io.CsvMapReader;
+import com.conveyal.gtfs.loader.Table;
+
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.StringReader;
+import java.io.IOException;
+import java.nio.file.*;
+import java.util.*;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.ArrayUtils;
+import org.supercsv.io.CsvMapWriter;
+import org.supercsv.prefs.CsvPreference;
+
+
+/**
+ * This feed transformation will attempt to preserve any custom fields from an entered csv in the final GTFS output.
+ */
+public class PreserveCustomFieldsTransformation extends ZipTransformation {
+    /** no-arg constructor for de/serialization */
+    public PreserveCustomFieldsTransformation() {}
+
+    public static PreserveCustomFieldsTransformation create(String csvData, String table) {
+        PreserveCustomFieldsTransformation transformation = new PreserveCustomFieldsTransformation();
+        transformation.csvData = csvData;
+        transformation.table = table;
+        return transformation;
+    }
+
+    @Override
+    public void validateParameters(MonitorableJob.Status status) {
+        if (csvData == null) {
+            status.fail("CSV data must not be null (delete table not yet supported)");
+        }
+    }
+
+    /**
+     * This method creates a hash map of the GTFS table keys to the custom CSV values for efficient lookup of custom values.
+     * The hash map key is the key values of the GTFS table (e.g. stop_id for stops) concatenated by an underscore.
+     * The hash map value is the CsvMapReader (mapping of column to row value).
+     */
+    private static HashMap<String, Map<String, String>> createCsvHashMap(CsvMapReader reader, String[] headers, List<String> tablePrimaryKeys) throws IOException {
+        HashMap<String, Map<String, String>> lookup = new HashMap<>();
+        Map<String, String> nextLine;
+        while ((nextLine = reader.read(headers)) != null) {
+            List<String> customCsvKeyValues = tablePrimaryKeys.stream().map(nextLine::get).collect(Collectors.toList());
+            String hashKey = StringUtils.join(customCsvKeyValues, "_");
+            lookup.put(hashKey, nextLine);
+        }
+        return lookup;
+    }
+
+    @Override
+    public void transform(FeedTransformZipTarget zipTarget, MonitorableJob.Status status) throws Exception{
+        String tableName = table + ".txt";
+        Path targetZipPath = Paths.get(zipTarget.gtfsFile.getAbsolutePath());
+        Optional<Table> streamResult = Arrays.stream(Table.tablesInOrder)
+                .filter(t -> t.name.equals(table))
+                .findFirst();
+
+        if (!streamResult.isPresent()) {throw new Exception(String.format("could not find specTable for table %s", table));}
```

Comment:
Reformat as
```
if (...) {
  throw...
}
```

Reply:
Updated in https://github.com/ibi-group/datatools-server/pull/536/commits/8fecae1d4f2ffc96a05a3fb7d88e5de1b7b6475a